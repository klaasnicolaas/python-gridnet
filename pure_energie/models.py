"""Models for Pure Energie Meter."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass
class SmartMeter:
    """Object representing an SmartMeter response from Pure Energie Meter."""

    power_flow: int
    energy_consumption_total: float
    energy_production_total: float

    @staticmethod
    def from_dict(data: dict[str, Any]) -> SmartMeter:
        """Return SmartMeter object from the Pure Energie Meter API response.

        Args:
            data: The data from the Pure Energie Meter API.

        Returns:
            A SmartMeter object.
        """
        data = data["elec"]

        def convert(value):
            """Convert the unit of measurement.

            Args:
                value: input value.

            Returns:
                Value in kWh rounded with 1 decimal.
            """
            value = value / 1000
            return round(value, 1)

        return SmartMeter(
            power_flow=data["power"]["now"].get("value"),
            energy_consumption_total=convert(data["import"]["now"].get("value")),
            energy_production_total=convert(data["export"]["now"].get("value")),
        )


@dataclass
class Device:
    """Object representing an Device response from Pure Energie Meter."""

    pem_id: str
    model: str
    batch: str
    firmware: str
    hardware: int
    manufacturer: str

    @staticmethod
    def from_dict(data: dict[str, Any]) -> Device:
        """Return Device object from the Pure Energie Meter API response.

        Args:
            data: The data from the Pure Energie Meter API.

        Returns:
            A Device object.
        """

        return Device(
            pem_id=data.get("id"),
            model=data.get("model"),
            batch=data.get("batch"),
            firmware=data.get("fw"),
            hardware=data.get("hw"),
            manufacturer=data.get("mf"),
        )
