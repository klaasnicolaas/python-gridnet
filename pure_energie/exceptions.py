"""Exceptions for Pure Energie Meter."""


class PureEnergieMeterError(Exception):
    """Generic Pure Energie Meter exception."""


class PureEnergieMeterConnectionError(PureEnergieMeterError):
    """Pure Energie Meter connection exception."""
