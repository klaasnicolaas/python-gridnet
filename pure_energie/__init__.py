"""Asynchronous Python client for the Pure Energie Meter API."""

from .exceptions import PureEnergieMeterConnectionError, PureEnergieMeterError
from .models import Device, SmartMeter
from .pure_energie import PureEnergie

__all__ = [
    "Device",
    "PureEnergie",
    "PureEnergieMeterError",
    "PureEnergieMeterConnectionError",
    "SmartMeter",
]
