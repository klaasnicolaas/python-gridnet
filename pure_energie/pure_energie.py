"""Asynchronous Python client for Pure Energie Meter API."""
from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Any, Mapping

import async_timeout
from aiohttp.client import ClientError, ClientResponseError, ClientSession
from yarl import URL

from .exceptions import PureEnergieMeterConnectionError
from .models import Device, SmartMeter


@dataclass
class PureEnergie:
    """Main class for handling connections with the Pure Energie Meter API."""

    def __init__(
        self, host: str, request_timeout: int = 10, session: ClientSession | None = None
    ) -> None:
        """Initialize connection with the Pure Energie Meter API.

        Args:
            host: Hostname or IP address of Pure Energie Meter device.
            request_timeout: An integer with the request timeout in seconds.
            session: Optional, shared, aiohttp client session.
        """
        self._session = session
        self._close_session = False

        self.host = host
        self.request_timeout = request_timeout

    async def request(
        self,
        uri: str,
        *,
        params: Mapping[str, str] | None = None,
    ) -> dict[str, Any]:
        """Handle a request to a Pure Energie device.

        Args:
            uri: Request URI, without '/', for example, 'status'
            params: Extra options to improve or limit the response.

        Returns:
            A Python dictionary (text) with the response from
            the Pure Energie device.

        Raises:
            PureEnergieMeterConnectionError: An error occurred while
                communicating with the Pure Energie device.
        """
        url = URL.build(scheme="http", host=self.host, path="/").join(URL(uri))

        headers = {
            "Accept": "application/json, text/plain, */*",
        }

        if self._session is None:
            self._session = ClientSession()
            self._close_session = True

        try:
            with async_timeout.timeout(self.request_timeout):
                response = await self._session.request(
                    "GET",
                    url,
                    params=params,
                    headers=headers,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            raise PureEnergieMeterConnectionError(
                "Timeout occurred while connecting to Pure Energie Meter device"
            ) from exception
        except (ClientError, ClientResponseError) as exception:
            raise PureEnergieMeterConnectionError(
                "Error occurred while communicating with the Pure Energie Meter device"
            ) from exception

        return await response.text()

    async def device(self) -> Device:
        """Get the latest values from the Pure Energie Meter.

        Returns:
            A Device data object from the Pure Energie device API.
        """
        data = await self.request("info")
        return Device.from_dict(data)

    async def smartmeter(self) -> SmartMeter:
        """Get the latest values from the Pure Energie Meter.

        Returns:
            A SmartMeter data object from the Pure Energie device API.
        """
        data = await self.request("meter/now")
        return SmartMeter.from_dict(data)

    async def close(self) -> None:
        """Close open client session."""
        if self._session and self._close_session:
            await self._session.close()

    async def __aenter__(self) -> PureEnergie:
        """Async enter.

        Returns:
            The Pure Energie Meter object.
        """
        return self

    async def __aexit__(self, *_exc_info) -> None:
        """Async exit.

        Args:
            _exc_info: Exec type.
        """
        await self.close()
