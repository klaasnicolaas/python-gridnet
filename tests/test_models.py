"""Test the models."""
import aiohttp
import pytest

from pure_energie import Device, PureEnergie, SmartMeter

from . import load_fixtures


@pytest.mark.asyncio
async def test_device(aresponses):
    """Test request from a Pure Energie device - Device object."""
    aresponses.add(
        "example.com",
        "/info",
        "GET",
        aresponses.Response(
            text=load_fixtures("device.json"),
            status=200,
        ),
    )

    async with aiohttp.ClientSession() as session:
        client = PureEnergie(host="example.com", session=session)
        device: Device = await client.device()
        assert device
        assert device.pem_id == "84df:0c11:9999:3795"
        assert device.model == "SBWF3102"
        assert device.firmware == "1.6.16"
        assert device.manufacturer == "NET2GRID"


@pytest.mark.asyncio
async def test_smartmeter(aresponses):
    """Test request from a Pure Energie device - SmartMeter object."""
    aresponses.add(
        "example.com",
        "/meter/now",
        "GET",
        aresponses.Response(
            text=load_fixtures("smartmeter.json"),
            status=200,
        ),
    )

    async with aiohttp.ClientSession() as session:
        client = PureEnergie(host="example.com", session=session)
        smartmeter: SmartMeter = await client.smartmeter()
        assert smartmeter
        assert smartmeter.power_flow == 338
        assert smartmeter.energy_consumption_total == 17762.1
        assert smartmeter.energy_production_total == 21214.6
