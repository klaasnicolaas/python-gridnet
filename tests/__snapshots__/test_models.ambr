# serializer version: 1
# name: test_device
  Device(n2g_id='84df:0c11:9999:3795', model='SBWF3102', batch='SBP-HMX-210318', firmware='1.6.16', hardware=1, manufacturer='NET2GRID')
# ---
# name: test_smartbridge
  SmartBridge(power_flow=338, energy_consumption_total=17762.1, energy_production_total=21214.6)
# ---
